### PR TITLE
DEV: consolidate chat channel notification settings

### PR DESCRIPTION
**Pull request by dbattersby**
### What is this change?

On the chat channel settings page, we want to show a single **Send push notifications** setting instead of the current Desktop notifications and Mobile push notifications settings.

Users can update the Send push notifications setting to receive push notifications:

- For all activity
- Only for mentions
- Never

For existing users, use the Mobile push notifications setting value for the new Send push notifications setting.

For new users, the Send push notifications setting should default to Only for mentions (which represents the current default).

### How it looks

Screenshot 2024-10-04 at 12 23 54 PM

/t/137659

### CODE REVIEW

File: plugins/chat/db/migrate/20241003122030_add_notification_level_to_user_chat_channel_memberships.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+class AddNotificationLevelToUserChatChannelMemberships < ActiveRecord::Migration[7.1]
+  def up
+    remove_index :user_chat_channel_memberships, name: "user_chat_channel_memberships_index"
+    add_column :user_chat_channel_memberships,
+               :notification_level,
+               :integer,
+               default: 1,
+               null: false
+
+    execute <<~SQL
```
Comment (dbattersby) [inline plugins/chat/db/migrate/20241003122030_add_notification_level_to_user_chat_channel_memberships.rb:11]: I am wondering if we need to do this update in batches?

**Review by markvanlan — APPROVED** [review]
This all looks good, it's pretty much just a swap of logic except for the migration. 

I don't believe we need any batching for the update query in the migration... Looking at previous migrations, the batch size is typically 500, 1000, or 10,000. The 500 batch size I'm looking at - https://github.com/discourse/discourse/blob/d959bfdc6198ae441336377fff10ed09a5f55d60/db/migrate/20191211170000_add_hashed_api_key.rb#L22

Is looping through row for each batch and executing a SQL cmd. We have far more instances of bulk updating records that _don't_ have batching and this table is constrained by # of users and # of chat channels. 

TLDR; I don't think we need batching.

**Comment by dbattersby** [comment]
After looking into some stats around the number of rows in the `user chat channel memberships` table I've opted to move the update logic into a post migration and apply batch processing to prevent table locking. I've also opted to remove and add the indexes concurrently as I think it's probably safer for very large sites.

File: plugins/chat/db/migrate/20241003122030_add_notification_level_to_user_chat_channel_memberships.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+class AddNotificationLevelToUserChatChannelMemberships < ActiveRecord::Migration[7.1]
+  disable_ddl_transaction!
+
+  def up
+    execute <<~SQL
+      DROP INDEX CONCURRENTLY IF EXISTS user_chat_channel_memberships_index
+    SQL
+
+    add_column :user_chat_channel_memberships, :notification_level, :integer, default: 1
```
[the remark this answers is not in the document]
Comment (dbattersby) [reply plugins/chat/db/migrate/20241003122030_add_notification_level_to_user_chat_channel_memberships.rb:10]: Thank you for helping sanity check this, there is quite a lot happening with these migrations.

I was thinking about moving the index and reapplying in a separate migration, I think it can be applied safely after everything else is complete. The initial migration only needs to focus on adding the new column.

I'm wondering is there any benefit to using post migrations here?

[the remark this answers is not in the document]
Comment (dbattersby) [reply plugins/chat/db/migrate/20241003122030_add_notification_level_to_user_chat_channel_memberships.rb]: I was thinking it might have been the combo of setting both default value and null false that done it before but better not to take the risk. I have updated the steps and moved this into a single migration, let me know what you think.

File: plugins/chat/db/migrate/20241003122030_add_notification_level_to_user_chat_channel_memberships.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+class AddNotificationLevelToUserChatChannelMemberships < ActiveRecord::Migration[7.1]
+  disable_ddl_transaction!
+
+  def up
+    execute <<~SQL
+      DROP INDEX CONCURRENTLY IF EXISTS user_chat_channel_memberships_index
+    SQL
+
+    add_column :user_chat_channel_memberships, :notification_level, :integer, null: true
```
Comment (tgxworld) [inline plugins/chat/db/migrate/20241003122030_add_notification_level_to_user_chat_channel_memberships.rb:10]: @dbattersby @markvanlan @nattsw We should not be adding a column or [changing a column's defaults](https://github.com/discourse/discourse/pull/29080/files#diff-d4b35a813304e060b02a2a2b55c5aadca4d2034ad9d97838d35d3d380a8673a5R21-R22) in a migration where `disable_ddl_transaction!` is used. Migrations with transactions disabled need to be idempotent so that we can run it again even if the migration fails.
